snuba/optimize_scheduler.py

```diff
@@ -0,0 +1,114 @@
+import re
+from dataclasses import dataclass
+from datetime import datetime, timedelta
+from typing import MutableSequence, Sequence
+
+from snuba import settings
+
+
+class OptimizedSchedulerTimeout(Exception):
+    """
+    The exception raised when the optimize scheduler is called to get the
+    next schedule after the final optimize job cutoff time.
+    """
+
+    pass
+
+
+@dataclass(frozen=True)
+class OptimizationSchedule:
+    partitions: Sequence[Sequence[str]]
+    cutoff_time: datetime
+
+
+class OptimizeScheduler:
+    """
+    The optimized scheduler provides a mechanism for scheduling optimizations
+    based on the parallelism desired.
+
+    If there is no parallelism desired, then it returns a single schedule
+    with the partitions provided.
+
+    If there is parallelism desired, then there is logic which determines
+    when parallelism can kick in and when it has to end. This is required
+    to avoid having too much load on the database.
+
+    If the scheduler is called to get next schedule after the last optimization
+    cutoff time then OptimizedSchedulerTimeout exception is raised.
+    """
+
+    def __init__(self, parallel: int) -> None:
+        self.__parallel = parallel
+        self.__last_midnight = (datetime.now() + timedelta(minutes=10)).replace(
+            hour=0, minute=0, second=0, microsecond=0
+        )
+        self.__parallel_start_time = (
+            self.__last_midnight + settings.PARALLEL_OPTIMIZE_JOB_START_TIME
+        )
+        self.__parallel_end_time = (
+            self.__last_midnight + settings.PARALLEL_OPTIMIZE_JOB_END_TIME
+        )
+        self.__full_job_end_time = (
+            self.__last_midnight + settings.OPTIMIZE_JOB_CUTOFF_TIME
+        )
+
+    @staticmethod
+    def subdivide_partitions(
+        partitions: Sequence[str], number_of_subdivisions: int
+    ) -> Sequence[Sequence[str]]:
+        """
+        Subdivide a list of partitions into number_of_subdivisions lists
+        of partitions so that optimizations can be executed in parallel.
+
+        We sort the partitions so that the more recent partitions are
+        optimized first since more recent partitions might have more
+        replacements.
+        """
+
+        def sort_partitions(partition_name: str) -> str:
+            date_regex = re.compile("\d{4}-(0[1-9]|1[0-2])-(0[1-9]|[12][0-9]|3[01])")
```

Comment:
I know this code was copied from another place, but it's bad practice to compile regexes in a loop. I would move this at least out of the `sort_partitions` function and maybe even into the init of the class to avoid this happening more than once.

Reply:
Good callout. Moved the regex compilation out.